### ISSUE TITLE
Get IP Address Upstream 

### ISSUE THREAD

**Issue by mjaverto** [jwilder/nginx-proxy#164]
Great project, I hope you can help me with a small piece and I can contribute back for others in our scenario. 

We are running self-contained spring boot jars that spin up a Tomcat server on 8080 per container. We need to reverse proxy back to the respective container based on NGINX locations. This project seems to be perfect for what we need but the only issue I run into is getting the container address, {{ $address.IP }}.

The error I get is on the upstream line: 

    2015/05/11 23:32:09 [emerg] 15#15: unexpected "{" in /etc/nginx/conf.d/my-custom.conf:29

as seen here when DOCKER_CONTAINER is an env correctly set for the container.

    {{ range $host, $containers := groupByMulti $ "Env.DOCKER_CONTAINER" "," }}

        upstream {{ $host }} {
            {{ range $container := $containers }}
                {{ range $address := $container.Addresses }}
                    # {{$container.Name}}
                    server {{ $address.IP }}:8080;
                {{ end }}
            {{ end }}
        }

    {{ end }}

I've tried editing many different ways and it seems like {{ $host }} is returning the env variable but I can't seem to get 

            {{ range $container := $containers }}
                {{ range $address := $container.Addresses }}
                    # {{$container.Name}}
                    server {{ $address.IP }}:8080;
                {{ end }}
            {{ end }}

working even if I remove the upstream {{ $host }} {} block.

Essentially I'm trying to get the upstream to look like this:

    upstream my-service-name {

        server 172.17.0.7:8080;

    }

And so it all makes sense, I have our locations hardcoded which works because the proxy pass is the same as Env.DOCKER_CONTAINER

    location /api/v1/sessions {

            proxy_pass         http://my-service-name;
            proxy_redirect     off;
            proxy_pass_request_headers      on;
            proxy_set_header   Host $host;
            proxy_set_header   X-Real-IP $remote_addr;
            proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header   X-Forwarded-Host $server_name;
            proxy_set_header   Origin $http_origin;
    }

Thanks in advance.
Docker version 1.6.0, build bdbc177

**Comment by mjaverto** [comment]
I actually think I see the problem now is related to:

https://github.com/jwilder/docker-gen/issues/79

for

        {{ range $container := $containers }}
            {{ range $address := $container.Addresses }}
                # {{$container.Name}}
                server {{ $address.IP }}:8080;
            {{ end }}
        {{ end }}

but I'm still trying to understand why I have the error with '{'

    2015/05/11 23:32:09 [emerg] 15#15: unexpected "{" in /etc/nginx/conf.d/my-custom.conf:29

when running:

    docker run --name=nginx-proxy --privileged=true -d -p 80:80 -v /home/mjaverto/dockerapps/nginx/nginx-proxy/nginx-dynamic.tmpl:/etc/nginx/conf.d/my-custom.conf:ro -v /var/run/docker.sock:/tmp/docker.sock jwilder/nginx-proxy

**Comment by md5** [comment]
The problem is that the files under `conf.d` are read directly by Nginx. They are not processed as templates. There is an open issue requesting the ability to include custom templates in #71. I believe that changes would be needed in `docker-gen` to support it.

As for jwilder/docker-gen#79, it's not related, though it is something you'd probably want to use for your use case if custom templates actually were supported.

**Comment by md5** [comment]
The discussion starting at this comment is also relevant: https://github.com/jwilder/nginx-proxy/pull/106#issuecomment-75466718

**Comment by mjaverto** [comment]
@md5 thanks for clarifying and the help. I was a little confused by that file in the root of the repo but I see now it works with Docker-gen.

**Comment by mjaverto** [comment]
So with fresh eyes this morning I figured everything out my problems for anyone else who may find this.

1. the template file I was passing into the container on start should have been mounted to /app/nginx.tmpl rather than into the nginx conf /etc/nginx/conf.d/my_proxy.conf dir. 

1. I had to expose some ports on my micro service containers so even though they could accept connections on 8080 without any ports exposed to the host machine, I still had to expose a few ports just to have docker-gen pick up their addresses.

My new docker run command is

    docker run --name=nginx-proxy -d -p 80:80 -v /home/user/dockerapps/nginx/nginx-proxy/nginx-dynamic.tmpl:/app/nginx.tmpl -v /var/run/docker.sock:/tmp/docker.sock jwilder/nginx-proxy

This is now working as needed for us and we are able to have multiple nginx locations reverse proxied to the one or many containers.
